### PR TITLE
AE-1212 update large borrows governance agent to use thresholds from contract getter functions

### PR DESCRIPTION
**Pull request by Nomad-AE**
- Updated the code so that the minimum votes required to propose or reach quorum will now be pulled from on-chain via the Compound Governor Alpha contract. 
- Fixed a bug where non-zero decimals would not evaluate properly.

### CODE REVIEW

File: large-borrows-governance/agent-config.json [hunk 1]
```diff
@@ -4,14 +4,13 @@
   "protocolAbbreviation": "COMP",
   "COMPAddress": "0xc00e94Cb662C3520282E6f5717214004A7f26888",
   "cCOMPAddress": "0x70e36f6BF80a52b3B46b3aF8e106CC0ed743E8e4",
+  "governorAddress": "0xc0dA01a04C3f3E0be433606045bB7017A7323E38",
```
Comment (echoAE) [inline large-borrows-governance/agent-config.json:7]:
```suggestion
  "governorAddress": "0xc0Da02939E1441F497fd74F78cE7Decb17B66529",
```

File: large-borrows-governance/abi/Governor.json [hunk 2]
```diff
@@ -0,0 +1,722 @@
+[
```
Comment (echoAE) [inline large-borrows-governance/abi/Governor.json:1]: I believe this is the ABI for an older Compound Governance smart contract (GovernorAlpha).  I _believe_ the ABI you want is present in this repository at `governance/abi/GovernorBravo.json`.

File: large-borrows-governance/src/agent.spec.js [hunk 3]
```diff
@@ -1,6 +1,25 @@
+const BigNumber = require('bignumber.js');
+
+// proposalThreshold and quorumVotes will be pulled from Governor Alpha contract
```
Comment (echoAE) [inline large-borrows-governance/src/agent.spec.js:3]:
```suggestion
// proposalThreshold and quorumVotes will be pulled from Governor Bravo contract
```

File: large-borrows-governance/src/agent.spec.js [hunk 4]
```diff
@@ -1,6 +1,25 @@
...
+// Note: the Proposal threshold will be less than the Quorum threshold
+const mockMinProposal = 1100;
+const mockMinQuorum = 4400;
+const mockDecimals = 2;
+
+// Convert to bignumber and 10^x
+let decimals = mockDecimals;
+decimals = new BigNumber(decimals.toString());
```
Comment (echoAE) [inline large-borrows-governance/src/agent.spec.js:11]:
```suggestion
let decimals = new BigNumber(mockDecimals);
```

File: large-borrows-governance/src/agent.spec.js [hunk 5]
```diff
@@ -1,6 +1,25 @@
...
+decimals = new BigNumber(10).pow(decimals);
+
+// convert to bignumber.js and divide by COMP decimals
+const minQuorumVotes = new BigNumber(mockMinQuorum.toString()).div(decimals);
+const minProposalVotes = new BigNumber(mockMinProposal.toString()).div(decimals);
```
Comment (echoAE) [inline large-borrows-governance/src/agent.spec.js:16]:
```suggestion
const minQuorumVotes = new BigNumber(mockMinQuorum).div(decimals);
const minProposalVotes = new BigNumber(mockMinProposal).div(decimals);
```

File: large-borrows-governance/src/agent.spec.js [hunk 6]
```diff
@@ -244,15 +270,17 @@ describe('handleTransaction', () => {
 
   it('returns two findings when a borrow event is emitted from the cCOMP token contract and the address that borrowed COMP exceeds the proposal and vote quorum thresholds', async () => {
     const {
-      minAmountCOMP: proposalMinCOMP, type: proposalType, severity: proposalSeverity,
+      type: proposalType, severity: proposalSeverity,
     } = borrowLevels.proposal;
     const {
-      minAmountCOMP: votingMinCOMP, type: votingQuorumType, severity: votingQuorumSeverity,
+      type: votingQuorumType, severity: votingQuorumSeverity,
     } = borrowLevels.votingQuorum;
```
Comment (echoAE) [inline large-borrows-governance/src/agent.spec.js:276]:
```suggestion
    const {
      proposal: {
        type: proposalType,
        severity: proposalSeverity,
      },
      votingQuorum: {
        type: votingQuorumType,
        severity: votingQuorumSeverity,
      },
    } = borrowLevels;
```

File: large-borrows-governance/src/agent.spec.js [hunk 7]
```diff
@@ -244,15 +270,17 @@ describe('handleTransaction', () => {
...
 
     // set the balanceOf to a value that is greater than the minimum COMP threshold for the proposal
     // and vote quorum governance interactions
-    mockERC20Contract.balanceOf.mockResolvedValue(votingMinCOMP + 1);
+    // Mock: minAmountCOMP + 1, and then with the rest of the decimals added back on.
+    mockERC20Contract.balanceOf.mockResolvedValue(BigNumber.sum(minQuorumVotes, 1)
+      .multipliedBy(decimals));
```
Comment (echoAE) [inline large-borrows-governance/src/agent.spec.js:282]:
```suggestion
    // Mock: minAmountCOMP + 1, and then with the rest of the decimals added back on.
    const currCOMPOwned = BigNumber.sum(minQuorumVotes, 1);
    mockERC20Contract.balanceOf.mockResolvedValue(currCOMPOwned.multipliedBy(decimals));
```

File: large-borrows-governance/src/agent.js [hunk 8]
```diff
@@ -95,34 +113,38 @@ function provideHandleTransaction(data) {
...
+              protocolAbbreviation,
+              type,
+              severity,
+              levelName,
+              borrowerAddress,
+              minAmountCOMP.toString(),
+              userCOMPBalance.toString(),
+            );
+          }
         }
```
Comment (arbitrary-CodeBeholder) [inline large-borrows-governance/src/agent.js:148]:
```suggestion
        const minAmountCOMP = voteMinimums[levelName];
        if ( minAmountCOMP !== undefined && userCOMPBalance.gte(minAmountCOMP ) {
          // a governance threshold has been crossed, generate an alert
          return createAlert(
            developerAbbreviation,
            protocolName,
            protocolAbbreviation,
            type,
            severity,
            levelName,
            borrowerAddress,
            minAmountCOMP.toString(),
            userCOMPBalance.toString(),
          );
        }
```

File: large-borrows-governance/src/agent.spec.js [hunk 9]
```diff
@@ -288,8 +316,8 @@ describe('handleTransaction', () => {
       metadata: {
         borrowerAddress: mockBorrowerAddress,
         governanceLevel: 'proposal',
-        minCOMPNeeded: proposalMinCOMP.toString(),
-        currCOMPOwned: (votingMinCOMP + 1).toString(),
+        minCOMPNeeded: minProposalVotes.toString(),
+        currCOMPOwned: (BigNumber.sum(minQuorumVotes, 1)).toString(),
```
Comment (echoAE) [inline large-borrows-governance/src/agent.spec.js:319]:
```suggestion
        currCOMPOwned: currCOMPOwned.toString(),
```

File: large-borrows-governance/src/agent.spec.js [hunk 10]
```diff
@@ -304,8 +332,8 @@ describe('handleTransaction', () => {
       metadata: {
         borrowerAddress: mockBorrowerAddress,
         governanceLevel: 'votingQuorum',
-        minCOMPNeeded: votingMinCOMP.toString(),
-        currCOMPOwned: (votingMinCOMP + 1).toString(),
+        minCOMPNeeded: minQuorumVotes.toString(),
+        currCOMPOwned: (BigNumber.sum(minQuorumVotes, 1)).toString(),
```
Comment (echoAE) [inline large-borrows-governance/src/agent.spec.js:335]:
```suggestion
        currCOMPOwned: currCOMPOwned.toString(),
```

**Review by echoAE — CHANGES_REQUESTED** [review]
Switch from GovernorAlpha address and ABI to GovernorBravo address and ABI.

Also, a few other minor changes.

File: large-borrows-governance/src/agent.spec.js [hunk 11]
```diff
@@ -199,8 +222,11 @@ describe('handleTransaction', () => {
   it('returns a finding when a borrow event is emitted from the cCOMP token contract and the address that borrowed COMP exceeds the proposal threshold', async () => {
     // set the balanceOf to a value that is greater than the minimum COMP threshold for the proposal
     // governance action
-    const { minAmountCOMP, type, severity } = borrowLevels.proposal;
-    mockERC20Contract.balanceOf.mockResolvedValue(minAmountCOMP + 1);
+    const { type, severity } = borrowLevels.proposal;
+    const minAmountCOMP = minProposalVotes;
+    // Mock: minAmountCOMP + 1, and then with the rest of the decimals added back on.
+    mockERC20Contract.balanceOf.mockResolvedValue(BigNumber.sum(minAmountCOMP, 1)
+      .multipliedBy(decimals));
```
[the remark this answers is not in the document]
Comment (Nomad-AE) [reply large-borrows-governance/src/agent.spec.js]: Added in commit 47e475a088c41501b9359dff228780c15bb0a43b

File: large-borrows-governance/src/agent.spec.js [hunk 12]
```diff
@@ -234,7 +260,7 @@ describe('handleTransaction', () => {
         borrowerAddress: mockBorrowerAddress,
         governanceLevel: 'proposal',
         minCOMPNeeded: minAmountCOMP.toString(),
-        currCOMPOwned: (minAmountCOMP + 1).toString(),
+        currCOMPOwned: (BigNumber.sum(minAmountCOMP, 1)).toString(),
```
[the remark this answers is not in the document]
Comment (Nomad-AE) [reply large-borrows-governance/src/agent.spec.js]: fixed in commit 47e475a088c41501b9359dff228780c15bb0a43b

File: large-borrows-governance/src/agent.spec.js (hunk 7 above)
Comment (Nomad-AE) [reply large-borrows-governance/src/agent.spec.js]: fixed in commit 47e475a088c41501b9359dff228780c15bb0a43b

File: large-borrows-governance/src/agent.js [hunk 13]
```diff
@@ -79,7 +96,8 @@ function provideInitialize(data) {
     data.borrowEvent = borrowEvent;
     data.borrowLevels = borrowLevels;
     data.contract = compContract;
-    data.decimalsExp = new BigNumber(10).pow(compDecimals);
+    data.decimalsExp = compDecimals;
+    data.voteMinimums = voteMinimums;
```
[the remark this answers is not in the document]
Comment (Nomad-AE) [reply large-borrows-governance/src/agent.js]: fixed in commit 47e475a088c41501b9359dff228780c15bb0a43b

File: large-borrows-governance/src/agent.spec.js (hunk 6 above)
Comment (Nomad-AE) [reply large-borrows-governance/src/agent.spec.js]: fixed in commit fd3e78c72eeba6a7d7018e8bd6685ba2b0dd06c1

File: large-borrows-governance/abi/Governor.json (hunk 2 above)
Comment (Nomad-AE) [reply large-borrows-governance/abi/Governor.json]: Resolved in commit 15e0b461bc512b459d4c97b6da942603d4ac1121

**Review by echoAE — APPROVED** [review]

**Pull request merged**